File: Queen2.java
public class Queen2 {
    private static int n = 15;
    private static int count = 0;
    private static boolean[] shu, pie, na;

    public static void main(String[] args) {
        shu = new boolean[n];
        pie = new boolean[2 * n - 1];
        na = new boolean[2 * n - 1];
        count = 0;
        long tic = System.currentTimeMillis();
        DFS(0);
        long toc = System.currentTimeMillis();
        System.out.println("Total solutions: " + count);
        System.out.println("Elapsed time: " + (toc - tic) + " ms");
    }

    private static void DFS(int row) {
        for (int col = 0; col < n; col++) {
            int i = row + col, j = n - 1 + col - row;
            if (shu[col] || pie[i] || na[j]) {
                continue;
            }
            if (row == n - 1) {
                count++;
            }
            else {
                shu[col] = pie[i] = na[j] = true;
                DFS(row + 1);
                shu[col] = pie[i] = na[j] = false;
            }
        }
    }
}

File: python/Queen2.py
import time

n = 13
shu = [False] * n
pie = [False] * (2 * n - 1)
na = [False] * (2 * n - 1)
count = 0

def DFS(row):
    global count
    for col in range(n):
        j = row + col; k = n - 1 - row + col
        if shu[col] or pie[j] or na[k]:
            continue
        if row == n - 1:
            count += 1
        else:
            shu[col] = pie[j] = na[k] = True
            DFS(row + 1)
            shu[col] = pie[j] = na[k] = False

tic = time.time()
DFS(0)
toc = time.time()
print "Total solutions: %d" % count
print "Elapsed time: %f seconds" % (toc - tic)
